### ISSUE TITLE
Add OSLC support

### ISSUE THREAD

**Issue by kuwv** [gaphor/gaphor#2223]
### Is your feature request related to a problem?

MBSE development tools need the capability to integrate with each other. I would like to be able to connect to an OSLC Architecture Management Provider using Gaphor as an OSLC consumer.

### Describe the solution you'd like

Open Services for Lifecycle Collaboration is interoperability framework for integrating MBSE stack components. These components can publish and consume data using a provider/consumer model. It would be nice if Gaphor was able to read and publish diagrams in this regard.

### Describe alternatives you've considered

Currently, one would have to integrate their own OSLC compatible plugins. It may be fine to just include a OSLC compatible template for plugins with a how-to.

### Additional context

https://open-services.net/
https://open-services.net/specifications/
https://archive.open-services.net/wiki/architecture-management/OSLC-Architecture-Management-Specification-Version-3.0/
https://oslcfest.org/presentations/PyOSLC_SDK.pdf
https://github.com/cslab/pyoslc/tree/master

**Comment by danyeaw** [comment]
Hi @kuwv, thanks for the suggestion. Could you detail out the use case you would like to see supported when you say that MBSE tools should be integrated with each other?

For example, with other modeling tools, I have seen them able to connect to a OSLC Provider (like a requirements tool) and then pull in data from that other tool. Then you can track linkage to that data in your model.

Depending on what you are trying to do, another option might be to develop a Model Development Kit for https://www.openmbee.org in order to sync data from Gaphor in to the Model Management System.

**Comment by danyeaw** [comment]
Do you have a threat modeling tool in mind that you would like to see integrated?

Gaphor is extensible, so it is possible that we could develop a plugin that would allow for this use case. However, developing a generic integration framework, even around something like OSLC, would be really difficult if we aren't targeting for an integration with another specific tool.

**Comment by amolenaar** [comment]
>  However, developing a generic integration framework, even around something like OSLC, would be really difficult if we aren't targeting for an integration with another specific tool.

I agree. Despite standards, there are always incompatibilities.

For my understanding: OSLC is mostly about serving data, right? Not so much updating it.

I think this can be a separate project even (using Gaphor as a library). An OSLC server is a web server, so it has no place in a GUI tool IMHO. Instead, one could create a little (flask/fastapi?) app that reads the model and serves it via HTTP API's.

**Comment by danyeaw** [comment]
I think the use case was the opposite. What you are describing is making Gaphor an OSLC Provider by serving up a Gaphor model to others. I think this feature request is to allow Gaphor to connect to other OSLC Providers to pull in data to a Gaphor model.

**Comment by kuwv** [comment]
@danyeaw - yup, I understand it would require time and I wouldn't suggest doing more - if any at all - than you are comfortable. I just figured I'd share my insights to allow you to decide.

I submitted issues with Threat Dragon to coordinate the same effort:
https://github.com/OWASP/threat-dragon/issues/639
https://github.com/OWASP/threat-dragon/issues/641

Also, a separate thread here: https://github.com/networkx/networkx/discussions/6632

@amolenaar - I updated the solution statement above to help answer your question. This would be a OSLC consumer. Essentially, there are several modeling clients that I know that act as OSLC consumers: MagicDraw, Capella (w/plugins) and System/Enterprise Architect.

I do really like the idea of using Graphor as a library and was going to open another issue depending on how receptive you were to this idea.

**Comment by amolenaar** [comment]
Thanks @kuwv, indeed I misinterpreted that. Being a consumer would make more sense in the context of Gaphor :smiley: 

> I do really like the idea of using Gaphor as a library and was going to open another issue depending on how receptive you were to this idea.

Gaphor *is* already a library :smiley:. Every (well written) python project has that capability. We made extra effort to separate all GUI related logic from the core application. That's how it's used to support [Sphinx](https://docs.gaphor.org/en/latest/sphinx.html) and [Jupyter notebooks](https://docs.gaphor.org/en/latest/scripting.html), for example.

In the end we also went as far to write a [pytest plugin](https://github.com/jwbargsten/pytest-archon) (with a colleague of mine), to ensure you can use Gaphor properly without GUI.

> A WebUI is also a cool idea but I think that would require the library so that the IDE would still function also.

I've been thinking on this idea on and off. I suppose a web viewer (no editing) would be feasible.

**Comment by danyeaw** [comment]
Since Threat Dragon would have to first build an OSLC provider, should we wait to see if that materializes before starting to build a client for Gaphor?

**Comment by kuwv** [comment]
There's other services that can be used. The OSLC provides an example with bigzilla w/ their plugin. Also, No Magic provides the API to Teamwork Cloud https://osmc.nomagic.com/

**Comment by amolenaar** [comment]
Since there's no free and open source OSLC applications, we cannot build an implementation. I'll close this issue. If someone is interested in building such an integration, that's much appreciated. It would probably be a separate project, though and not part of Gaphor (core).
